src/renderer/canvas/label.ts

```diff
@@ -58,40 +72,160 @@ export class Label {
-      label.position.x - width / 2,
-      label.position.y - baselineHeight - margin + i * lineHeight,
-      width,
-      height,
-    );
+
+    let width = context.measureText(line).width + 2 * margin;
+    let x = label.position.x - width / 2;
+    let y = label.position.y - baselineHeight - margin + i * lineHeight;
+    let height = label.fontSize + 2 * margin;
+
+    if (label.properties.fontBackgroundPadding) {
+      width += label.properties.fontBackgroundPadding * 2;
+      height += label.properties.fontBackgroundPadding * 2;
+      x -= label.properties.fontBackgroundPadding;
+
+      if (type === 'node') {
+        y += label.properties.fontBackgroundPadding * 2 * i;
+      } else {
+        if (i === 0) {
+          y -= label.properties.fontBackgroundPadding;
+        } else {
+          y += label.properties.fontBackgroundPadding;
+        }
+      }
+    }
+
+    borderPoints.push({ x, y, width, height });
+  }
+
+  return borderPoints;
+};
+
+const calculateBorderRadius = (borderRadius: number, width: number, height: number) => {
+  // ensure that the radius isn't too large for x
+  if (width - 2 * borderRadius < 0) {
+    borderRadius = width / 2;
+  }
+
+  // ensure that the radius isn't too large for y
+  if (height - 2 * borderRadius < 0) {
+    borderRadius = height / 2;
+  }
+
+  return borderRadius;
+};
+
+const drawTextBackground = (context: CanvasRenderingContext2D, borderPoints: BorderPoint[], label: Label) => {
+  if (!label.properties.fontBackgroundColor) {
+    return;
+  }
+
+  const backgroundColor = label.properties.fontBackgroundColor ?? DEFAULT_FONT_BACKGROUND_COLOR;
+  const fontBorderRadius = label.properties.fontBackgroundBorderRadius ?? 0;
+
+  context.fillStyle = backgroundColor.toString();
+
+  for (let i = 0; i < borderPoints.length; i++) {
+    const { x, y, width, height } = borderPoints[i];
+    const borderRadius = calculateBorderRadius(fontBorderRadius, width, height);
+    drawRoundRect(context, x, y, width, height, borderRadius, i === 0, true);
+    context.fill();
   }
 };
 
-const drawText = (context: CanvasRenderingContext2D, label: Label) => {
+const drawTextBackgroundBorder = (context: CanvasRenderingContext2D, borderPoints: BorderPoint[], label: Label) => {
+  if (!label.properties.fontBackgroundBorderWidth) {
+    return;
+  }
+
+  const borderWidth = label.properties.fontBackgroundBorderWidth;
+  const borderColor = label.properties.fontBackgroundBorderColor ?? DEFAULT_FONT_BORDER_COLOR;
+  const fontBorderRadius = label.properties.fontBackgroundBorderRadius ?? 0;
+  context.lineWidth = borderWidth;
+  context.strokeStyle = borderColor.toString();
+
+  context.beginPath();
+
+  if (borderPoints.length === 1) {
+    const { x, y, height, width } = borderPoints[0];
+    const borderRadius = calculateBorderRadius(fontBorderRadius, width, height);
+    drawRoundRect(context, x, y, width, height, borderRadius);
+  } else {
+    for (let i = 0; i < borderPoints.length; i++) {
```

Comment:
I am not 100% sure why you need this logic here. I feel like all this can be done using 3 functions which are: `drawRoundRect` + `context.fill()` and `context.stroke()`.

Here is the idea:

* There are two stages: draw background and draw foreground (the actual text)
* Stage 1: Background
  * baseline height in case if text baseline is `TOP` should incorporate the padding (so it doesn't overlay what is above it, e.g. node circle)
  * for each line of text get the bounding boxes (what you called BorderPoints) - check the `context.font` for `context.measureText` (it is in another comment)
  * for each line call `drawRoundRect` - what is cool is that it will make a path and it will not fill nor stroke it - which is great
  * after all lines are called with `drawRoundRect` we can just do 2 calls: `context.stroke()` (if backgroundBorderWidth > 0)  and `context.fill()` - this fill will override the overlapping strokes.
  
You can test this last piece with this little sample:

```html
<!doctype html>
<html lang="en-US">
  <head>
    <meta charset="utf-8" />
    <title>Canvas</title>
    <style>
      canvas {
        border: 1px solid black;
      }
    </style>
  </head>
  <body>
    <canvas id="tutorial" width="550" height="550"></canvas>
    <script>
      function draw() {
        const canvas = document.getElementById("tutorial");
        if (canvas.getContext) {
          const ctx = canvas.getContext("2d");
          ctx.fillStyle = 'yellow';
          ctx.lineWidth = 1;
          ctx.strokeStyle = 'black';
          ctx.rect(100, 100, 20, 20);
          ctx.rect(110, 110, 20, 20);
          ctx.stroke();
          ctx.fill();
        }
      }
      draw();
    </script>
  </body>
</html>
```

* Stage 2: Foreground
  * Everything should stay the same as before, the only difference is the starting `y` when the text baseline is `TOP` (in the case of the nodes) because of the padding - so the baseline height needs to incorporate the padding.

What do you think?

Reply:
I think there will be one issue here if there is no backgroundColor i.e. `label.properties.fontBackgroundColor` is not set, in this case with the above method, we will get overlapping borders around the label.

![image](https://github.com/memgraph/orb/assets/39862387/c14b10ed-69ab-46b8-b8eb-a6deb87647f4)

So, the border has to be drawn independently of the background.

Please let me know if my understanding is not correct.